## Analysis of US Alternative Fuel Stations

In [1]:
import pandas as pd

In [2]:
#1 import numpy
import numpy as np

In [3]:
#2 link to the alt_fuel_stations_Jan 2024.csv
df = pd.read_csv('C:\\Users\\bukep\\Desktop\\Data_Science\\Alternative Fuels Data center -US Eenergy\\alt_fuel_stations_Jan_14_2024.csv')

C:\Users\bukep\AppData\Local\Temp\ipykernel_41676\2764453876.py:2: DtypeWarning: Columns (6,16,20,31,33,39,40,41,43,46,52,55,57,58,60,62,65,67,69,71) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('C:\\Users\\bukep\\Desktop\\Data_Science\\Alternative Fuels Data center -US Eenergy\\alt_fuel_stations_Jan_14_2024.csv')


In [4]:
df.head()

Fuel Type Code                                  Station Name  \
0            CNG          Spire - Montgomery Operations Center   
1            CNG  Metropolitan Atlanta Rapid Transit Authority   
2            CNG                         United Parcel Service   
3            CNG                    Arkansas Oklahoma Gas Corp   
4            CNG    Clean Energy - Logan International Airport   

         Street Address                            Intersection Directions  \
0      2951 Chestnut St                                                NaN   
1   2424 Piedmont Rd NE                                                NaN   
2  270 Marvin Miller Dr                                                NaN   
3     2100 S Waldron Rd                                                NaN   
4   1000 Cottage St Ext  From Route 1, take the first exit after Callah...   

          City State    ZIP  Plus4 Station Phone Status Code  ...  \
0   Montgomery    AL  36107    NaN           NaN           E  ...   
1      Atlanta    GA  30324    NaN           NaN           E  ...   
2      Atlanta    GA  30336    NaN           NaN           E  ...   
3   Fort Smith    AR  72903    NaN  479-783-3181           E  ...   
4  East Boston    MA   2128    NaN  866-809-4869           E  ...   

  Restricted Access RD Blends RD Blends (French) RD Blended with Biodiesel  \
0               NaN       NaN                NaN                       NaN   
1               NaN       NaN                NaN                       NaN   
2               NaN       NaN                NaN                       NaN   
3             False       NaN                NaN                       NaN   
4             False       NaN                NaN                       NaN   

  RD Maximum Biodiesel Level NPS Unit Name  \
0                        NaN           NaN   
1                        NaN           NaN   
2                        NaN           NaN   
3                        NaN           NaN   
4                        NaN           NaN   

  CNG Station Sells Renewable Natural Gas  \
0                                   False   
1                                     NaN   
2                                    True   
3                                   False   
4                                    True   

   LNG Station Sells Renewable Natural Gas  Maximum Vehicle Class  \
0                                      NaN                     MD   
1                                      NaN                     LD   
2                                      NaN                     HD   
3                                      NaN                     MD   
4                                      NaN                     MD   

   EV Workplace Charging  
0                    NaN  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN  

[5 rows x 74 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94369 entries, 0 to 94368
Data columns (total 74 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Fuel Type Code                           94369 non-null  object 
 1   Station Name                             94368 non-null  object 
 2   Street Address                           94335 non-null  object 
 3   Intersection Directions                  5089 non-null   object 
 4   City                                     94365 non-null  object 
 5   State                                    94354 non-null  object 
 6   ZIP                                      94369 non-null  object 
 7   Plus4                                    0 non-null      float64
 8   Station Phone                            87172 non-null  object 
 9   Status Code                              94369 non-null  object 
 10  Expected Date                            6521 

In [8]:
df.columns

Index(['Fuel Type Code', 'Station Name', 'Street Address',
       'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser N

In [5]:
#1 - Address all missing values
#count the nans
df.isnull().sum()

Fuel Type Code                                 0
Station Name                                   1
Street Address                                34
Intersection Directions                    89280
City                                           4
                                           ...  
NPS Unit Name                              94185
CNG Station Sells Renewable Natural Gas    93414
LNG Station Sells Renewable Natural Gas    94301
Maximum Vehicle Class                      72150
EV Workplace Charging                      12182
Length: 74, dtype: int64

In [11]:
#find nans 
df.isnull()

Fuel Type Code  Station Name  Street Address  Intersection Directions  \
0               False         False           False                     True   
1               False         False           False                     True   
2               False         False           False                     True   
3               False         False           False                     True   
4               False         False           False                    False   
...               ...           ...             ...                      ...   
94364           False         False           False                     True   
94365           False         False           False                     True   
94366           False         False           False                     True   
94367           False         False           False                     True   
94368           False         False           False                     True   

        City  State    ZIP  Plus4  Station Phone  Status Code  ...  \
0      False  False  False   True           True        False  ...   
1      False  False  False   True           True        False  ...   
2      False  False  False   True           True        False  ...   
3      False  False  False   True          False        False  ...   
4      False  False  False   True          False        False  ...   
...      ...    ...    ...    ...            ...          ...  ...   
94364  False  False  False   True          False        False  ...   
94365  False  False  False   True          False        False  ...   
94366  False  False  False   True          False        False  ...   
94367  False  False  False   True          False        False  ...   
94368  False  False  False   True          False        False  ...   

       Restricted Access  RD Blends  RD Blends (French)  \
0                   True       True                True   
1                   True       True                True   
2                   True       True                True   
3                  False       True                True   
4                  False       True                True   
...                  ...        ...                 ...   
94364               True       True                True   
94365               True       True                True   
94366               True       True                True   
94367               True       True                True   
94368               True       True                True   

       RD Blended with Biodiesel  RD Maximum Biodiesel Level  NPS Unit Name  \
0                           True                        True           True   
1                           True                        True           True   
2                           True                        True           True   
3                           True                        True           True   
4                           True                        True           True   
...                          ...                         ...            ...   
94364                       True                        True           True   
94365                       True                        True           True   
94366                       True                        True           True   
94367                       True                        True           True   
94368                       True                        True           True   

       CNG Station Sells Renewable Natural Gas  \
0                                        False   
1                                         True   
2                                        False   
3                                        False   
4                                        False   
...                                        ...   
94364                                     True   
94365                                     True   
94366                                     True   
94367                                     True 

In [8]:
df.describe()

Plus4  EV Level1 EVSE Num  EV Level2 EVSE Num  EV DC Fast Count  \
count    0.0          755.000000        71367.000000      12062.000000   
mean     NaN            4.200000            2.410736          3.789753   
std      NaN            8.169848            3.162759          4.697930   
min      NaN            1.000000            1.000000          1.000000   
25%      NaN            1.000000            2.000000          1.000000   
50%      NaN            2.000000            2.000000          2.000000   
75%      NaN            3.500000            2.000000          4.000000   
max      NaN           90.000000          338.000000         84.000000   

           Latitude     Longitude             ID  Federal Agency ID  \
count  94369.000000  94369.000000   94369.000000        1322.000000   
mean      39.165938    -95.183994  187443.964406          13.563540   
std        5.711686     19.411451   78795.768102           6.158342   
min      -41.354100   -164.848855      17.000000           2.000000   
25%       34.392261   -117.574477  144203.000000           8.000000   
50%       39.399690    -89.316809  187321.000000          14.000000   
75%       42.975710    -78.593763  238224.000000          16.000000   
max       64.852466    122.373276  325442.000000          33.000000   

       CNG Dispenser Num  CNG Total Compression Capacity  \
count        1029.000000                      622.000000   
mean            3.029155                      915.500000   
std             6.505245                     1075.692844   
min             0.000000                        2.000000   
25%             1.000000                      250.000000   
50%             2.000000                      700.000000   
75%             2.000000                     1200.000000   
max            80.000000                    10620.000000   

       CNG Storage Capacity  RD Blends (French)  RD Maximum Biodiesel Level  
count            319.000000                 0.0                  537.000000  
mean           46889.423197                 NaN                    5.370577  
std            56246.507440                 NaN                    4.391574  
min                0.000000                 NaN                    5.000000  
25%            29203.500000                 NaN                    5.000000  
50%            36000.000000                 NaN                    5.000000  
75%            58000.000000                 NaN                    5.000000  
max           593136.000000                 NaN                   99.000000

Question 1.	What is the number and percentage of the values of the “Fuel Type Code” column?

In [9]:
df['Fuel Type Code'].value_counts().sort_values(ascending=False)

Fuel Type Code
ELEC    82214
E85      4527
LPG      3528
BD       1702
CNG      1530
RD        610
HY        133
LNG       125
Name: count, dtype: int64

In [13]:
percentages = (df['Fuel Type Code'].value_counts(normalize=True) * 100).round(1).astype(str) + '%'
print(percentages)


Fuel Type Code
ELEC    87.1%
E85      4.8%
LPG      3.7%
BD       1.8%
CNG      1.6%
RD       0.6%
HY       0.1%
LNG      0.1%
Name: proportion, dtype: object


In [16]:
counts = df['Fuel Type Code'].value_counts()
percentages = df['Fuel Type Code'].value_counts(normalize=True)
result = pd.concat([counts, percentages], axis=1, keys=['count', 'percentage'])
print(result)

                count  percentage
Fuel Type Code                   
ELEC            82214    0.871197
E85              4527    0.047971
LPG              3528    0.037385
BD               1702    0.018036
CNG              1530    0.016213
RD                610    0.006464
HY                133    0.001409
LNG               125    0.001325


In [14]:
df['Fuel Type Code'].describe()

count     94369
unique        8
top        ELEC
freq      82214
Name: Fuel Type Code, dtype: object

In [14]:
fuel_type_codes = df['Fuel Type Code'].value_counts(normalize=True)

In [17]:
num_unique_values = len(fuel_type_codes)

In [18]:
percentage_values = (fuel_type_codes * 100).round(2)

In [19]:
print("Number of unique values in 'Fuel Type Code':", num_unique_values)
print("\nPercentage of each value in 'Fuel Type Code':")
for code, percentage in zip(fuel_type_codes.index, percentage_values):
    print(f"{code}: {percentage}%")

Number of unique values in 'Fuel Type Code': 8

Percentage of each value in 'Fuel Type Code':
ELEC: 87.12%
E85: 4.8%
LPG: 3.74%
BD: 1.8%
CNG: 1.62%
RD: 0.65%
HY: 0.14%
LNG: 0.13%


In [12]:
df.head(3)

Fuel Type Code                                  Station Name  \
0            CNG          Spire - Montgomery Operations Center   
1            CNG  Metropolitan Atlanta Rapid Transit Authority   
2            CNG                         United Parcel Service   

         Street Address Intersection Directions        City State    ZIP  \
0      2951 Chestnut St                     NaN  Montgomery    AL  36107   
1   2424 Piedmont Rd NE                     NaN     Atlanta    GA  30324   
2  270 Marvin Miller Dr                     NaN     Atlanta    GA  30336   

   Plus4 Station Phone Status Code  ... Restricted Access RD Blends  \
0    NaN           NaN           E  ...               NaN       NaN   
1    NaN           NaN           E  ...               NaN       NaN   
2    NaN           NaN           E  ...               NaN       NaN   

  RD Blends (French) RD Blended with Biodiesel RD Maximum Biodiesel Level  \
0                NaN                       NaN                        NaN   
1                NaN                       NaN                        NaN   
2                NaN                       NaN                        NaN   

  NPS Unit Name CNG Station Sells Renewable Natural Gas  \
0           NaN                                   False   
1           NaN                                     NaN   
2           NaN                                    True   

   LNG Station Sells Renewable Natural Gas  Maximum Vehicle Class  \
0                                      NaN                     MD   
1                                      NaN                     LD   
2                                      NaN                     HD   

   EV Workplace Charging  
0                    NaN  
1                    NaN  
2                    NaN  

[3 rows x 74 columns]

Number of Fuel Type code = 94369
Percentage of each value in 'Fuel Type Code':
ELEC: 87.12%
E85: 4.8%
LPG: 3.74%
BD: 1.8%
CNG: 1.62%
RD: 0.65%
HY: 0.14%
LNG: 0.13%

Question 2.	What is the number and percentage of the top 3 most common states? Make an “insight” about the location of alternative fuel centers. Where are they being deployed most?

In [131]:
print(df.columns.tolist())
top_states = df['State'].value_counts().head(3)
total_centers = len(df)
top_states_percentage = (top_states / total_centers * 100).round(2)

['Fuel Type Code', 'Station Name', 'Street Address', 'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4', 'Station Phone', 'Status Code', 'Expected Date', 'Groups With Access Code', 'Access Days Time', 'Cards Accepted', 'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num', 'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network', 'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude', 'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code', 'Federal Agency ID', 'Federal Agency Name', 'Open Date', 'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary', 'E85 Blender Pump', 'EV Connector Types', 'Country', 'Intersection Directions (French)', 'Access Days Time (French)', 'BD Blends (French)', 'Groups With Access Code (French)', 'Hydrogen Is Retail', 'Access Code', 'Access Detail Code', 'Federal Agency Code', 'Facility Type', 'CNG Dispenser Num', 'CNG On-Site Renewable Source', 'CNG Total Compression Capacity', 'CNG Storage Capacity', 'LNG On-S

In [23]:
print("Top 3 most common states for alternative fuel centers:")
print(top_states)
print("\nPercentage of total alternative fuel centers:")
print(f"{top_states}:{top_states_percentage}%")

Top 3 most common states for alternative fuel centers:
State
CA    19402
QC     4573
NY     4478
Name: count, dtype: int64

Percentage of total alternative fuel centers:
State
CA    19402
QC     4573
NY     4478
Name: count, dtype: int64:State
CA    20.56
QC     4.85
NY     4.75
Name: count, dtype: float64%


OR question 2 can be done this way and arrive with the same answer

In [18]:
# Get value counts for states
state_counts = df['State'].value_counts()

# Get the top 3 states
top_3_states = state_counts.head(3)

# Calculate percentages
total_stations = len(df)
percentages = (top_3_states / total_stations) * 100

# Create a DataFrame with counts and percentages
top_states_df = pd.DataFrame({
    'State': top_3_states.index,
    'Number of Stations': top_3_states.values,
    'Percentage': percentages.values
})

# Display the result
print(top_states_df)


  State  Number of Stations  Percentage
0    CA               19402   20.559718
1    QC                4573    4.845871
2    NY                4478    4.745202


OR

In [10]:
# Count the number of stations per state and get the top 3
state_counts = df['State'].value_counts().head(3)
total_stations = df['State'].count()

# Calculate the percentage of stations in these top 3 states
state_percentages = (state_counts / total_stations) * 100

# Display the results
print("Top 3 most common states for alternative fuel stations:")
print(state_counts)
print("\nPercentage of stations in the top 3 states:")
print(state_percentages)

# Insight about the location of alternative fuel centers
print("\nInsight:")
print("The majority of alternative fuel stations are concentrated in the top 3 states, indicating a focus on deploying these resources in specific regions. This could be driven by a variety of factors, including state policies, the availability of alternative fuels, demand for cleaner transportation options, or infrastructure readiness. These states likely represent areas where the transition to alternative fuels is.")


Top 3 most common states for alternative fuel stations:
State
CA    19402
QC     4573
NY     4478
Name: count, dtype: int64

Percentage of stations in the top 3 states:
State
CA    20.562986
QC     4.846641
NY     4.745957
Name: count, dtype: float64

Insight:
The majority of alternative fuel stations are concentrated in the top 3 states, indicating a focus on deploying these resources in specific regions. This could be driven by a variety of factors, including state policies, the availability of alternative fuels, demand for cleaner transportation options, or infrastructure readiness. These states likely represent areas where the transition to alternative fuels is


OR

In [9]:
# Get the top 3 most common states
top_states = df['State'].value_counts().nlargest(3)

# Calculate the percentage of each top state
total_rows = len(df)
top_states_percent = top_states / total_rows * 100

# Print the top 3 states and their percentage
print("Top 3 top_states:")
for i, state in enumerate(top_states.index):
    print(f"{i+1}. {state}: {top_states_percent[i]:.2f}%")

Top 3 top_states:
1. CA: 20.56%
2. QC: 4.85%
3. NY: 4.75%


C:\Users\bukep\AppData\Local\Temp\ipykernel_6700\2857369576.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"{i+1}. {state}: {top_states_percent[i]:.2f}%")


Question 3.	Look at the column called “EV_Network.” What are the top 3 most common EV charging networks?

In [130]:
print(df.columns.tolist())

['Fuel Type Code', 'Station Name', 'Street Address', 'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4', 'Station Phone', 'Status Code', 'Expected Date', 'Groups With Access Code', 'Access Days Time', 'Cards Accepted', 'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num', 'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network', 'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude', 'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code', 'Federal Agency ID', 'Federal Agency Name', 'Open Date', 'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary', 'E85 Blender Pump', 'EV Connector Types', 'Country', 'Intersection Directions (French)', 'Access Days Time (French)', 'BD Blends (French)', 'Groups With Access Code (French)', 'Hydrogen Is Retail', 'Access Code', 'Access Detail Code', 'Federal Agency Code', 'Facility Type', 'CNG Dispenser Num', 'CNG On-Site Renewable Source', 'CNG Total Compression Capacity', 'CNG Storage Capacity', 'LNG On-S

From dictionary entry with the following keys: 
{
    "Index(['Fuel Type'": "       'EV Level2 EVSE Num'",
    " 'Station Name'": 'EV DC Fast Count'",
 " 'Street Address'": " 'EV Other Info'",
    "": " 'EV Network'"
  }
  print(df['EV_Network'].value_counts().head(3))
  print(df['EV_Network'].value_counts(normalize=True).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).tail(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).tail(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).tail(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  

.nlargest(3): This function returns the top 3 rows of the resulting Series, sorted in descending order by count. In this case, it gives  the 3 most common EV charging networks based on their occurrence counts.

In [79]:
top_3_networks = df['EV Network'].value_counts().nlargest(3)

print(top_3_networks)


EV Network
ChargePoint Network    39821
Non-Networked          12355
Blink Network           6451
Name: count, dtype: int64


4.	I want to buy a Tesla. I live in Mississippi. How many charging stations are there in my state?

In [101]:
# Filter by state (Mississippi) and fuel type (Tesla)
mississippi_tesla_stations = df[(df['State'] == 'MS') & (df['EV Network'].str.contains('Tesla'))]

# Get the count of Tesla charging stations in Mississippi
num_tesla_stations_in_mississippi = len(mississippi_tesla_stations)

# Display the result
print(f"Number of Tesla charging stations in Mississippi: {num_tesla_stations_in_mississippi}")


Number of Tesla charging stations in Mississippi: 42


Question 5.	What percentage of stations are owned by the Federal government? 

In [120]:
# Display all column names in the DataFrame
print(df.columns.tolist())

['Fuel Type Code', 'Station Name', 'Street Address', 'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4', 'Station Phone', 'Status Code', 'Expected Date', 'Groups With Access Code', 'Access Days Time', 'Cards Accepted', 'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num', 'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network', 'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude', 'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code', 'Federal Agency ID', 'Federal Agency Name', 'Open Date', 'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary', 'E85 Blender Pump', 'EV Connector Types', 'Country', 'Intersection Directions (French)', 'Access Days Time (French)', 'BD Blends (French)', 'Groups With Access Code (French)', 'Hydrogen Is Retail', 'Access Code', 'Access Detail Code', 'Federal Agency Code', 'Facility Type', 'CNG Dispenser Num', 'CNG On-Site Renewable Source', 'CNG Total Compression Capacity', 'CNG Storage Capacity', 'LNG On-S

In [69]:
# Filter the DataFrame to only include rows with Federal government as the owner type
federal_gov_df = df[df['Owner Type Code'] == 'Federal Government']

# Calculate the percentage of Federal government-owned stations
federal_gov_percentage = (federal_gov_df.shape[0] / df.shape[0]) * 100

# Print the percentage of Federal government-owned stations
print(f"{federal_gov_percentage:.2f}% of alternative fuel stations are owned by the Federal government.")

0.00% of alternative fuel stations are owned by the Federal government.


Question 6.	What are the top 3 most popular facility types? Why do you think those facility types are the most popular?

In [50]:
# Count the number of occurrences of each facility type
facility_type_counts = df['Facility Type'].value_counts()

# Get the top 3 most common facility types
top_3_facility_types = facility_type_counts.head(3)

# Print the top 3 most common facility types
print("Top 3 most common facility types:")
print(top_3_facility_types)

Top 3 most common facility types:
Facility Type
CONVENIENCE_STORE    4912
HOTEL                3454
CAR_DEALER           3116
Name: count, dtype: int64


In [53]:
top_facility_types = df['Facility Type'].value_counts()
top_3_facility_types = facility_type_counts.head(3)
print(top_facility_types)

Facility Type
CONVENIENCE_STORE    4912
HOTEL                3454
CAR_DEALER           3116
GAS_STATION          2469
FUEL_RESELLER        1467
                     ... 
PHARMACY               12
WORKPLACE               6
MOTOR_POOL              5
STADIUM                 3
TNC                     1
Name: count, Length: 63, dtype: int64


In [31]:
# Get value counts for facility types
facility_counts = df['Facility Type'].value_counts()

# Get the top 3 facility types
top_3_facilities = facility_counts.head(3)

# Display the result
print("Top 3 Most Popular Facility Types:")
print(top_3_facilities)


Top 3 Most Popular Facility Types:
Facility Type
CONVENIENCE_STORE    4912
HOTEL                3454
CAR_DEALER           3116
Name: count, dtype: int64


Question 7. For Level 2 EV charging stations (column “EV Level2 EVSE Num”), how many Level 2 EV Charging stations have more than 10 chargers?

In [125]:
print(df.columns.tolist())

['Fuel Type Code', 'Station Name', 'Street Address', 'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4', 'Station Phone', 'Status Code', 'Expected Date', 'Groups With Access Code', 'Access Days Time', 'Cards Accepted', 'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num', 'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network', 'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude', 'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code', 'Federal Agency ID', 'Federal Agency Name', 'Open Date', 'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary', 'E85 Blender Pump', 'EV Connector Types', 'Country', 'Intersection Directions (French)', 'Access Days Time (French)', 'BD Blends (French)', 'Groups With Access Code (French)', 'Hydrogen Is Retail', 'Access Code', 'Access Detail Code', 'Federal Agency Code', 'Facility Type', 'CNG Dispenser Num', 'CNG On-Site Renewable Source', 'CNG Total Compression Capacity', 'CNG Storage Capacity', 'LNG On-S

In [54]:
# Filter the DataFrame to only include rows with Level 2 EV charging stations
level2_df = df[df['EV Level2 EVSE Num'] > 0]

# Count the number of Level 2 EV charging stations with more than 10 chargers
more_than_10_chargers = (level2_df['EV Level2 EVSE Num'] > 10).sum()

# Print the number of Level 2 EV charging stations with more than 10 chargers
print(f"There are {more_than_10_chargers} Level 2 EV charging stations with more than 10 chargers.")

There are 946 Level 2 EV charging stations with more than 10 chargers.


In [56]:
num_stations_with_more_than_10_chargers = df[(df['EV Level2 EVSE Num'] > 10)].count()
print(num_stations_with_more_than_10_chargers)

Fuel Type Code                             946
Station Name                               946
Street Address                             946
Intersection Directions                     99
City                                       946
                                          ... 
NPS Unit Name                                2
CNG Station Sells Renewable Natural Gas      0
LNG Station Sells Renewable Natural Gas      0
Maximum Vehicle Class                      351
EV Workplace Charging                      945
Length: 74, dtype: int64


Question 8. List 3 questions that you would want to ask if you are looking to open a new alternative fuel center.
Below are the questions, i would ask for alternative fuel centers;

Question A: What is the demand for alternative fuel in the area?
Question B: What is the distribution of alternative fuel types in the area?
Question C: What is the distribution of facility types in the area?

In [ ]:
# Question A: What is the demand for alternative fuel in the area?
# Count the number of alternative fuel stations in the area
num_stations = df[(df['State'] == 'your_state') & (df['City'] == 'your_city')].shape[0]

# Print the number of alternative fuel stations in the area
print(f"There are currently {num_stations} alternative fuel stations in {your_city}, {your_state}.")

In [32]:
#Question B: What is the distribution of alternative fuel types in the area?
# Count the number of occurrences of each alternative fuel type in the area
fuel_type_counts = df[(df['State'] == 'your_state') & (df['City'] == 'your_city')]['Fuel Type Code'].value_counts()

# Print the distribution of alternative fuel types in the area
print("Distribution of alternative fuel types in {your_city}, {your_state}:")
print(fuel_type_counts)

Distribution of alternative fuel types in {your_city}, {your_state}:
Series([], Name: count, dtype: int64)


In [33]:
# Question C: What is the distribution of facility types in the area?
# Count the number of occurrences of each facility type in the area
facility_type_counts = df[(df['State'] == 'your_state') & (df['City'] == 'your_city')]['Facility Type'].value_counts()

# Print the distribution of facility types in the area
print("Distribution of facility types in {your_city}, {your_state}:")
print(facility_type_counts)

Distribution of facility types in {your_city}, {your_state}:
Series([], Name: count, dtype: int64)


Question 9.	What is an “insight” that you could communicate to someone who is thinking about whether to invest in a new alternative fuel center

These are the insights that i would communicate to investor to be in an alternative fuel center:

The demand for alternative fuel is growing, and there is a trend towards increased adoption of electric vehicles (EVs) in particular. According to the International Energy Agency - see reference, the number of electric cars on the road worldwide surpassed 1 million in 2020, a 40% year-on-year increase. This trend is expected to continue, with the number of electric cars projected to reach 140 million by 2030.

Investing in an alternative fuel center now could position you to take advantage of this growing market. However, it is important to carefully consider the location and type of alternative fuel center, as well as the current state of alternative fuel infrastructure in the area. By analyzing data on alternative fuel stations, such as the data available from the US Department of Energy's Alternative Fuels Data Center, you can make informed decisions about the type and location of your new alternative fuel center.

Additionally, it is important to consider the potential environmental benefits of investing in alternative fuel. Alternative fuels, such as electricity, hydrogen, and biofuels, can produce lower greenhouse gas emissions than traditional fossil fuels, which can help reduce the carbon footprint of transportation. By investing in alternative fuel, you can contribute to a more sustainable future and potentially attract customers who prioritize environmental concerns.

Overall, investing in an alternative fuel center can be a smart business decision that aligns with growing market trends and environmental goals. However, it is important to carefully consider the location, type, and potential environmental benefits of your investment.

Based on the reference below;
1. International Energy Agency. (2021). Global EV Outlook 201: Achieving an Electric Vehicle Future. Retrieved https://www.iea.org/reports/global--outlook-2021
2. US Department of Energy's Alternative Fuels Data Center. Alternative Fueling Station Counts by State and County. Retrieved from https://afdc.energy.gov/data/stations/counts
3. US Department of Energy's Alternative Fuels Data Center. Alternative Fueling Station Locations by Fuel Type. Retrieved from https://afdc.energy.gov/data/stations/stationsbyfueltype.html
4. US Environmental Protection Agency. Greenhouse Gas Emissions from Transportation. Retrieved from https://www.epa.gov/greenvehicles/fast-facts-transportation-greenhouse-gas-emissions 

Question 10.	List an insight or fact about a column of your choosing. List the column name and give some piece of information about that column. 

In [67]:
# Choosing the state column
# check for missing values in the state column
print(df['State'].value_counts().head())
print("Missing values in the 'State' column: ", df['State'].isnull().sum())

State
CA    19402
QC     4573
NY     4478
ON     4379
TX     4168
Name: count, dtype: int64
Missing values in the 'State' column:  15


In [62]:
# Counting the number of unique values in the "State" column
print("Number of unique states: ", df['State'].nunique())

Number of unique states:  65


In [68]:
# calculate the percentage of rows for each state
state_counts = df['State'].value_counts(normalize=True) 
print("Percentage of rows for each state:\n", state_counts)

Percentage of rows for each state:
 State
CA    0.205630
QC    0.048466
NY    0.047460
ON    0.046410
TX    0.044174
        ...   
AK    0.000699
YT    0.000435
PR    0.000233
NT    0.000032
KA    0.000011
Name: proportion, Length: 65, dtype: float64
